### PR TITLE
feature #24: Add Find and replace for graphs

### PR DESCRIPTION
**Pull request by phmz**
Add: new icon in the graph toolbar
Add: new menu in FSGraph
Add: find and replace dialog
Update: GraphFrame,UnitexFrame and InternalFrameManager now implement
find and replace
Update: MyCursors has a new icon for find and replace
Update: new method isStandaloneBox in GenericGraphBox
Update: new method setTextBox in GenericGraphicalZone

### CODE REVIEW

**Comment by martinec** [comment]
Hi @phmz, before merging your pull request please consider to open a new issue [here](https://github.com/UnitexGramLab/gramlab-ide/issues/new). This is important to let us know what you are trying to achieve and how we can help.

**Comment by eric-laporte** [comment]
Approved

**Comment by aleksandrachasch** [comment]
hey, @phmz, I have reviewed your PR and I've got several suggestions:

1. In the following scenario the feature does not work as expected:
     <b>What steps will reproduce the problem?</b>

    1.Launch Unitex.jar
    2.Select French as working language
    3.FSGraph > Create > New
    4.Draw a graph as on the screenshot:
    
screenshot1


   
    
  5.Select "Regex". 
    6.In "Find" field put `.*`
    7.In "Replace" field put `working!`

<b>What is the expected output?</b>
    The text inside the box is `working!`

<b>What do you see instead?</b>
   The text inside the box is `working!working!`
    
screenshot1

Is the scenario supposed to be like this, or it is an error?

2. A very small suggestion. If selecting `Replace all`, if number of occurrences is 0 or 1, the Message dialog will be as follow:
screenshot1    
screenshot1
The message text seems to be grammatically incorrect in these cases. May be it could be nice to rephrase the text, e.g. `Replaced occurrences: 0`?

3. Select FSGraph > Find and replace > in all graphs
    
screenshot1
In order to proceed to the 'Find and Replace' dialog box, it is needed to press 'Red Cross' button on the top right corner on `Opened graph` frame. It made me a little bit confusing, as I thought if I click Red Cross I will quit all the process of 'Replace and Find'.
To my mind, it would be great to add "OK" or "Continue" button to the `Opened graph` frame so that users will not misunderstand the purpose of this dialog.

**Comment by phmz** [comment]
hi @aleksandrachasch, as I mentioned you cannot replace the content of a standalone box the reason being standalone box are used for comment (page 96 of the Unitex User Manual).

I followed your suggestion and changed the message:
capture
 
I have also added a new button to close the Opened graph dialog
capture2

**Comment by aleksandrachasch** [comment]
Approved ! :+1:

**Comment by mdamis** [comment]
Approved

File: unitex/src/fr/umlv/unitex/frames/UnitexFindAndReplaceDialog.java [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+        gbc.gridy = 1;
+        gbc.gridwidth = 2;
+        gbc.weightx = 1.0;
+        gbc.weighty = 1.0;
+        gbc.anchor = GridBagConstraints.WEST;
+        gbc.fill = GridBagConstraints.HORIZONTAL;
+        gbc.insets = new Insets(0, 0, 0, 20);
+        panel1.add(replaceTextField, gbc);
+        JLabel findLabel = new JLabel();
+        findLabel.setText("Find:");
```
Comment (martinec) [inline unitex/src/fr/umlv/unitex/frames/UnitexFindAndReplaceDialog.java]: I suggest : ``Find what:``

File: unitex/src/fr/umlv/unitex/frames/UnitexFindAndReplaceDialog.java [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+        gbc = new GridBagConstraints();
+        gbc.gridx = 0;
+        gbc.gridy = 0;
+        gbc.weighty = 1.0;
+        gbc.anchor = GridBagConstraints.WEST;
+        panel1.add(findLabel, gbc);
+        JLabel replaceLabel = new JLabel();
+        replaceLabel.setText("Replace:");
```
Comment (martinec) [inline unitex/src/fr/umlv/unitex/frames/UnitexFindAndReplaceDialog.java]: I suggest : ``Replace with:``

File: unitex/src/fr/umlv/unitex/frames/UnitexFrame.java [hunk 3]
```diff
@@ -939,6 +941,24 @@ public void actionPerformed(ActionEvent e) {
 		};
 		graphMenu.add(new JMenuItem(printAll));
 		graphMenu.addSeparator();
+		final JMenu findAndReplaceMenu = new JMenu("Find and replace");
+
+		final Action findAndReplaceAll = new AbstractAction("in all graphs") {
```
Comment (martinec) [inline unitex/src/fr/umlv/unitex/frames/UnitexFrame.java]: I suggest: ``In all graphs``

File: unitex/src/fr/umlv/unitex/frames/UnitexFrame.java [hunk 4]
```diff
@@ -939,6 +941,24 @@ public void actionPerformed(ActionEvent e) {
...
+			@Override
+			public void actionPerformed(ActionEvent e) {
+				findAndReplaceAll();
+			}
+		};
+		final Action findAndReplace = new AbstractAction("in current graph") {
```
Comment (martinec) [inline unitex/src/fr/umlv/unitex/frames/UnitexFrame.java]: I suggest: ``In current graph``

**Comment by martinec** [comment]
Here attached a resized version (24x24 over a 32x32 canvas) of the find icon.

findandreplaceicon

**Comment by martinec** [comment]
How about integrate the "Opened Graph" information into the Find and Replace dialog box? My suggestion is as follows:

**Current dialog boxes:**

opened

findar

**Proposed dialog box:**

find and replace

- **Find what**: a dropdown list of recently used search terms.

- **Replace with**:  a dropdown list of recently used replace terms.

-  **Graphs:**: a dropdown list of open graphs. For the working example:

```
All open graphs
Current graph
new graph
A7.grf
```

When a user clicks the dropdown to select an item, if the item selected is a graph (i.e. neither ``All open graphs`` nor  ``Current graph``), the window containing the graph is show up. 

- **Use regular expressions**: a check-box to use regular expressions in the search term.

- **Case sensitive**: a check-box for case sensitive search. 

- **Match only a whole line**: a check-box to force a whole line match (See User's Manual section 5.2.1). e.g. if the search term is ``<N+Hum>`` it matches ``<N+Hum>+<PRE>`` but not ``<N+Hum><<^A>>+<PRE>``

- **Find previous**: a button to look for a box containing the search term before the current box.

- **Find next**: a button to look for a box containing the search term after the current box.

- **Replace**: a button to replace the matched occurrence with the replace term.

- **Replace all**: a button to replace all matched occurrences with the replace term.

- **Close**: a button to close the dialog.

- **Status bar**: a status bar to show information about the number of matched boxes while searching: "Found ``n`` boxes which match for ``search term``".

**Comment by martinec** [comment]
In the following scenario, the find and search feature  does not work as expected:

**What steps will reproduce the problem?**

- Launch Unitex.jar
- Select English as working language
- FSGraph > Open > test_find_and_replace.grf (here attached)

find_and_replace_test

- FSGraph > Find > In current graph
- Find for ``box``
- Click Next (**3 boxes matches**)
- Replace with ``foo``
- Click Replace all

**What is the expected behavior?**

``box`` is replaced by ``foo`` in all (3) matched boxes

**What do you see instead?**

``box`` is replaced by  ``foo`` only in 1 matched box

**``test_find_and_replace.grf``**
```
#Unigraph
SIZE 1188 840
FONT Inconsolata:  10
OFONT Inconsolata:B 12
BCOLOR 16777215
FCOLOR 0
ACOLOR 13487565
SCOLOR 16711680
CCOLOR 255
DBOXES y
DFRAME y
DDATE y
DFILE y
DDIR n
DRIG n
DRST n
FITS 100
PORIENT L
#
5
"<E>" 70 200 1 4 
"" 347 200 0 
"/This is a comment box" 137 139 0 
"This is a box without connections" 107 170 0 
"This is a connected box" 130 200 1 1 
```

**Comment by phmz** [comment]
I followed @martinec suggestions.

The new dialog looks like this:

capture3

- **Graphs**: a dropdown list of open graphs, it will update if a graph is opened or closed.

- **Ignore comment boxes**: a checkbox to ignore the comment boxes.

- **Status bar**: update as soon as the user input a character in **Find what**.

I also have added tooltips to display the graph path.

capture2

**Comment by aleksandrachasch** [comment]
Approved ! To my mind, everything works great now

**Comment by martinec** [comment]
In the following scenario, the find and search feature  does not work as expected:

**What steps will reproduce the problem?**

- Launch Unitex.jar
- Select English as working language
- FSGraph > Open >English/Graphs/Preprocessing/Sentence/Interjection.grf
- FSGraph > Find and replace
- Find for ``<A>``

**What is the expected behavior?**

``No match found with: <A>``

**What do you see instead?**

``Found 3 boxes which match with: <A>``

**Comment by martinec** [comment]
In the following scenario, the find and search feature  does not work as expected:

**What steps will reproduce the problem?**

- Launch Unitex.jar
- Select English as working language
- FSGraph > Open >English/Graphs/Preprocessing/Sentence/Interjection.grf
- FSGraph > Find and replace
- Find for ``Ah``

**What is the expected behavior?**

Both ``Replace`` and ``Replace All`` are disabled if: 1) ``Replace with:`` is empty or 2) there are not matches.

**What do you see instead?**

Both ``Replace`` and ``Replace All`` are enabled.  The replace function does not work correctly.

**Comment by martinec** [comment]
In the following scenario, the find and search feature  does not work as expected:

**What steps will reproduce the problem?**

- Launch Unitex.jar
- Select English as working language
- FSGraph > Open >English/Graphs/Preprocessing/Sentence/Interjection.grf
- FSGraph > Find and replace
- Find what: ``<E>`` (6 boxes which match)
- Replace with: ``<E>/output``
- Replace All

**What is the expected behavior?**

 ``Replace 6 boxes with: <E>/output``

**What do you see instead?**

``No match found with: <E>``

**Comment by martinec** [comment]
In the following scenario, the find and search feature  does not work as expected:

**What steps will reproduce the problem?**

- Launch Unitex.jar
- Select English as working language
- FSGraph > Open >English/Graphs/Preprocessing/Sentence/Interjection.grf
- FSGraph > Find and replace
- Find what: ``damn`` (on box ``Amen+damn+darn+Sit down``)
- Replace with: ``<``
- Replace All

**What is the expected behavior?**

An error message: ``Error in the box content you are replacing. No closing >``

**What do you see instead?**

An exception is thrown:

```bash
fr.umlv.unitex.exceptions.NoClosingSupException
	at fr.umlv.unitex.graphrendering.GraphBox.tokenizeTextWithoutTransduction(Unknown Source)
```

This is the result of replacing ``Amen+damn+darn+Sit down``  by  ``Amen+<+darn+Sit down`` which is an invalid expression.

**Comment by phmz** [comment]
In order to fix [this](https://github.com/UnitexGramLab/gramlab-ide/pull/23#issuecomment-222723839), empty boxes (containing only \<E>) will be ignored by both **Find** and **Replace**.

In [this case](https://github.com/UnitexGramLab/gramlab-ide/pull/23#issuecomment-222721635) where **Replace what** is empty **Replace** and **Replace All** will replace the **Find what** string with an empty string.

Finally, if one or boxes are not valid after a replace an error message will be displayed in the **Status bar**

findandreplace5

File: unitex/src/fr/umlv/unitex/graphrendering/GraphBox.java [hunk 5]
```diff
@@ -32,487 +32,607 @@
...
 
-	private static boolean isAPlusChar(char c) {
-		return (c == '+' || c == '\uff0b');
-	}
+  private static boolean isAPlusChar(char c) {
+    return (c == '+' || c == '\uff0b');
+  }
 
-	private void tokenizeTextWithoutTransduction(String s)
-			throws BackSlashAtEndOfLineException, MissingGraphNameException,
```
Comment (martinec) [inline unitex/src/fr/umlv/unitex/graphrendering/GraphBox.java]: Do not transform formatting of previous sources

File: unitex/src/fr/umlv/unitex/graphrendering/GraphBox.java [hunk 6]
```diff
@@ -514,5 +514,128 @@ void drawOtherStandalone(Graphics2D g, DrawGraphParams params) {
...
+   * This method checks if the String s is valid or not.
+   *
+   * @param s the string to check
+   * @throws BackSlashAtEndOfLineException if the String ends with a backslash.
+   * @throws MissingGraphNameException if the String starts with a semicolon but is missing a graph name.
+   * @throws NoClosingQuoteException if the String contains a not closed quotation.
+   * @throws NoClosingSupException if the String contains a not closed chevron.
+   * @throws NoClosingRoundBracketException if the String contains a not closed brace.
+   */
+  public void checkString(String s)
```
Comment (martinec) [inline unitex/src/fr/umlv/unitex/graphrendering/GraphBox.java]: This function seems to be highly similar (96%) to [tokenizeTextWithoutTransduction](../../blob/9762cb327c5c04c9148abc413d69a579b53173ea/unitex/src/fr/umlv/unitex/graphrendering/GraphBox.java#L79-193). What was the motivation behind to duplicate the code? Do you see some possibility to refactor it to reduce duplication?

File: unitex/src/fr/umlv/unitex/graphtools/FindAndReplace.java [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
+ *
+ */
+package fr.umlv.unitex.graphtools;
+
+import com.sun.corba.se.impl.orbutil.graph.Graph;
```
Comment (aleksandrachasch) [inline unitex/src/fr/umlv/unitex/graphtools/FindAndReplace.java]: Eclipse IDE signals that this package is never used

File: unitex/src/fr/umlv/unitex/frames/FindAndReplaceDialog.java [hunk 8]
```diff
@@ -0,0 +1,633 @@
...
+        tooltips.add(f.getGraph().getPath());
+      }
+    }
+    renderer.setTooltips(tooltips);
+    graphComboBox.setRenderer(renderer);
+    graphComboBox.setModel(model);
+  }
+
+  private void onClose() {
+    dispose();
```
Comment (aleksandrachasch) [inline unitex/src/fr/umlv/unitex/frames/FindAndReplaceDialog.java]: I've just checked that usually in UnitexFrames closing action for buttons is defined right in the `actionPerformed(ActionEvent e){  }` body. For example [here](https://github.com/UnitexGramLab/gramlab-ide/blob/master/unitex/src/fr/umlv/unitex/frames/ProcessInfoFrame.java) lines 139-141

Comment (phmz) [inline unitex/src/fr/umlv/unitex/frames/FindAndReplaceDialog.java]: onClose() is called this way too see [here](https://github.com/UnitexGramLab/gramlab-ide/pull/23/files#diff-58192b8da892411e52aae1010bd54ab2R100).

**Comment by alexis-neme** [comment]
Bug:
Test: Open NC-*.grf from French>inflection folder

Find what: <$2>
Graphs : All open graphs
Status bar: Found 14 boxes matches with <$2>

Button "Find Next" does not display the next graph with the box in green as expected.

**Comment by aleksandrachasch** [comment]
In the following scenario, the find and search feature does not work as expected:

**What steps will reproduce the problem?**

   * Launch Unitex.jar
   * Select French as working language
   * FSGraph > Open >French/Graphs/Normalization/Norm.grf
   * FSGraph > Find and replace
   * Find what: `Post`
   * Replace with: `foo`
   * Replace All
   * Close "Find and replace" dialog
   * Close the graph
   * FSGraph > Open > French/Graphs/Normalization/Contractions.grf
   * FSGraph > Find and replace
    
**What is the expected behavior?**

The statusBar is empty. The text in `Find what:` and `Replace with:` is selected.

**What do you see instead?**
The statusBar contains message from the previous search. Text in `Find what:` and `Replace with:` is not selected. 

default

**Comment by Mukarr** [comment]
Hey, @phmz I have a suggestion.
After clicking on `Replace` or `Replace All` once, if the user clicks on `Find Previous` or `Find Next` the status bar should be updated to number of matches found for given input.

**Comment by phmz** [comment]
@aleksandrachasch Thanks. Fixed it.
@Mukarr Thanks for the suggestion I have implemented it.

**Comment by mufeedalali** [comment]
Approved

**Comment by aleksandrachasch** [comment]
Approved

**Comment by martinec** [comment]
LGTM

**Pull request merged**